File: nodejs/test/e2e/client_api.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Client session management", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    async function waitFor(predicate: () => Promise<boolean>, timeoutMs = 10_000): Promise<void> {
        const deadline = Date.now() + timeoutMs;
        while (Date.now() < deadline) {
            if (await predicate()) return;
            await new Promise((resolve) => setTimeout(resolve, 50));
        }
        throw new Error(`Condition was not met within ${timeoutMs}ms`);
    }

    async function assertFailure(
        action: () => Promise<unknown>,
        expectedMessage: string
    ): Promise<void> {
        await expect(action()).rejects.toSatisfy((err: unknown) => {
            const text = err instanceof Error ? `${err.message}\n${err.stack ?? ""}` : String(err);
            expect(text.toLowerCase()).toContain(expectedMessage.toLowerCase());
            return true;
        });
    }

    it("should get null last session id before any sessions exist", async () => {
        await client.start();

        const result = await client.getLastSessionId();
        expect(result).toBeFalsy();
    });

    it("should delete session by id", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        const sessionId = session.sessionId;

        await session.sendAndWait({ prompt: "Say OK." });
        await waitFor(async () =>
            (await client.listSessions()).some((s) => s.sessionId === sessionId)
        );
        await session.disconnect();
        await client.deleteSession(sessionId);

        const metadata = await client.getSessionMetadata(sessionId);
        expect(metadata).toBeFalsy();
    }, 60_000);

    it("should report error when deleting unknown session id", async () => {
        await client.start();
        const unknownSessionId = "00000000-0000-0000-0000-000000000000";

        await assertFailure(
            () => client.deleteSession(unknownSessionId),
            `Failed to delete session ${unknownSessionId}`
        );
    });

    it("should track last session id after session created", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        await session.sendAndWait({ prompt: "Say OK." });
        const sessionId = session.sessionId;
        await session.disconnect();

        const lastId = await client.getLastSessionId();
        expect(lastId).toBe(sessionId);
    });

    it("should get null foreground session id in headless mode", async () => {
        await client.start();

        const sessionId = await client.getForegroundSessionId();
        expect(sessionId).toBeFalsy();
    });

    it("should report error when setting foreground session in headless mode", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });

        await assertFailure(
            () => client.setForegroundSessionId(session.sessionId),
            "Not running in TUI+server mode"
        );

        await session.disconnect();
    });
});

File: go/internal/e2e/client_api_e2e_test.go
package e2e

import (
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

// Mirrors dotnet/test/ClientSessionManagementTests.cs (snapshot category "client_api").
func TestClientApiE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	t.Run("should delete session by id", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		sessionID := session.SessionID

		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say OK."}); err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if err := session.Disconnect(); err != nil {
			t.Fatalf("Failed to disconnect session: %v", err)
		}

		if err := client.DeleteSession(t.Context(), sessionID); err != nil {
			t.Fatalf("Failed to delete session: %v", err)
		}

		metadata, err := client.GetSessionMetadata(t.Context(), sessionID)
		if err != nil {
			t.Fatalf("Failed to query session metadata: %v", err)
		}
		if metadata != nil {
			t.Errorf("Expected metadata to be nil after delete, got %+v", metadata)
		}
	})

	t.Run("should report error when deleting unknown session id", func(t *testing.T) {
		sessionID := "00000000-0000-0000-0000-000000000000"
		err := client.DeleteSession(t.Context(), sessionID)
		if err == nil {
			t.Fatal("Expected DeleteSession to fail for unknown id")
		}
		expectedMessage := "failed to delete session " + sessionID
		if !strings.Contains(strings.ToLower(err.Error()), expectedMessage) {
			t.Errorf("Expected error mentioning %q, got %v", expectedMessage, err)
		}
	})

	t.Run("should get null last session id before any sessions exist", func(t *testing.T) {
		// Use a fresh client with isolated COPILOT_HOME so other subtests don't pollute state.
		freshCtx := testharness.NewTestContext(t)
		freshClient := freshCtx.NewClient()
		t.Cleanup(func() { freshClient.ForceStop() })

		if err := freshClient.Start(t.Context()); err != nil {
			t.Fatalf("Failed to start fresh client: %v", err)
		}

		result, err := freshClient.GetLastSessionID(t.Context())
		if err != nil {
			t.Fatalf("Failed to get last session id: %v", err)
		}
		if result != nil {
			t.Errorf("Expected nil last session id on fresh client, got %q", *result)
		}
	})

	t.Run("should track last session id after session created", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		sessionID := session.SessionID

		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say OK."}); err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if err := session.Disconnect(); err != nil {
			t.Fatalf("Failed to disconnect session: %v", err)
		}

		lastID, err := client.GetLastSessionID(t.Context())
		if err != nil {
			t.Fatalf("Failed to get last session id: %v", err)
		}
		if lastID == nil || *lastID != sessionID {
			got := "<nil>"
			if lastID != nil {
				got = *lastID
			}
			t.Errorf("Expected last session id %q, got %q", sessionID, got)
		}
	})

	t.Run("should get null foreground session id in headless mode", func(t *testing.T) {
		sessionID, err := client.GetForegroundSessionID(t.Context())
		if err != nil {
			t.Fatalf("Failed to get foreground session id: %v", err)
		}
		if sessionID != nil {
			t.Errorf("Expected nil foreground session id in headless mode, got %q", *sessionID)
		}
	})

	t.Run("should report error when setting foreground session in headless mode", func(t *testing.T) {
		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		t.Cleanup(func() { session.Disconnect() })

		err = client.SetForegroundSessionID(t.Context(), session.SessionID)
		if err == nil {
			t.Fatal("Expected SetForegroundSessionID to fail in headless mode")
		}
		if !strings.Contains(err.Error(), "Not running in TUI+server mode") {
			t.Errorf("Expected error mentioning 'Not running in TUI+server mode', got %v", err)
		}
	})
}
